### PR TITLE
Generate BuildConfig

### PR DESCRIPTION
**Pull request by ralph-bergmann**
This commit adds a generator which generates a `BuildConfig` class located at `lib/build_config.g.dart`. 

The generator runs when `flutter run` or `flutter build` is called (or its equivalent IDE methods).

The `BuildConfig` class holds some information about the current build; it looks like this:

```
// GENERATED CODE - DO NOT MODIFY BY HAND
// Will be (re-)generated by flutter run or flutter build command.

class BuildConfig {
  static const String kBuildTime = '2018-05-31T00:57:39.780680';
  static const int kBuildTimeMillisecondsSinceEpoch = 1527721059780;
  static const bool kDebug = false;
  static const String kModeName = 'release';
  static const String kFlavor = null;
  static const String kVersionName = '1.0.2';
  static const int kVersionNumber = 3;
  static const Map<String, dynamic> kBuildArguments = {
    'build-name': '1.0.2',
    'build-number': '5',
    'help': false,
    'target': 'lib/main.dart',
    'debug': false,
    'profile': false,
    'release': false,
    'pub': true,
    'preview-dart-2': true,
    'track-widget-creation': false,
    'prefer-shared-library': false,
    'target-platform': 'android-arm'
  };
}
```
If the project is executed or build with native tools the generator will not (re-)generate the `BuildConfig` class.

### CODE REVIEW

**Comment by Hixie** [comment]
Can you elaborate on what this patch does?

cc @tvolkert

File: packages/flutter_tools/lib/src/buildconfig_generator.dart [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        ..type = refer('String')
+        ..assignment = new Code('"${buildInfo.modeName}"')))
+      ..fields.add(new Field((FieldBuilder b) => b
+        ..name = 'kFlavor'
+        ..static = true
+        ..modifier = FieldModifier.constant
+        ..type = refer('String')
+        ..assignment = new Code('${buildInfo.flavor != null ? "buildInfo.flavor" : null }')))
+
+      // TODO wait for https://github.com/flutter/flutter/pull/16857 to be merged
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/buildconfig_generator.dart]: (minor nit: our syntax for TODOs is `// TODO(username): description, https://link_to_issue`

File: packages/flutter_tools/templates/create/lib/main.dart.tmpl [hunk 2]
```diff
@@ -24,6 +25,7 @@ class MyApp extends StatelessWidget {
   // This widget is the root of your application.
   @override
   Widget build(BuildContext context) {
+    print('Flutter Demo runs in ${BuildConfig.kModeName} mode');
```
Comment (Hixie) [inline packages/flutter_tools/templates/create/lib/main.dart.tmpl]: i'm guessing this wasn't intentional?

File: packages/flutter_tools/templates/create/lib/build_config.g.dart.tmpl [hunk 3]
```diff
@@ -0,0 +1,13 @@
+// GENERATED CODE - DO NOT MODIFY BY HAND
+
+class BuildConfig {
+  static const bool kDebug = true;
```
Comment (Hixie) [inline packages/flutter_tools/templates/create/lib/build_config.g.dart.tmpl]: this presumably has to be generated at build time so shouldn't be in the template?

File: packages/flutter_tools/pubspec.yaml [hunk 4]
```diff
@@ -11,8 +11,10 @@ dependencies:
   archive: 1.0.33
   args: 1.4.2
   cli_util: 0.1.2+1
+  code_builder: 3.0.3
   coverage: 0.10.0
   crypto: 2.0.2+1
+  dart_style: 1.0.10
```
Comment (Hixie) [inline packages/flutter_tools/pubspec.yaml]: We're actually trying to reduce our dependencies currently so we probably will want to avoid adding these if we can help it.

**Comment by the4thfloor** [comment]
@Hixie 

When I call `flutter build apk --debug --flavor="foo"` the Flutter part knows nothing about it.

On Android, I have the `BuildConfig` class which knows all about the build, for example:
```
public final class BuildConfig {
  public static final boolean DEBUG = Boolean.parseBoolean("true");
  public static final String APPLICATION_ID = "eu.the4thfloor.msync.debug";
  public static final String BUILD_TYPE = "debug";
  public static final String FLAVOR = "";
  public static final int VERSION_CODE = 33;
  public static final String VERSION_NAME = "v1.0.15-debug (Build: d729bb8)";
}
```
This path creates a `lib/build_config.g.dart` file like this one:
```
// GENERATED CODE - DO NOT MODIFY BY HAND

class BuildConfig {
  static const bool kDebug = true;
  static const String kModeName = "debug";
  static const String kFlavor = null;
  static const String kVersionName = "1.0.0";
  static const int kVersionNumber = 1;
}
```
So it will be more comfortable on Dart side to do something depending on the build.

File: packages/flutter_tools/pubspec.yaml (hunk 4 above)
Comment (the4thfloor) [reply packages/flutter_tools/pubspec.yaml]: I can generate the code by hand but maybe it will not right formatted.

Comment (Hixie) [reply packages/flutter_tools/pubspec.yaml]: We don't actually use exact dartfmt style either so technically it might not be right formatted either way. :-)

File: packages/flutter_tools/lib/src/buildconfig_generator.dart [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  final StringBuffer output = new StringBuffer();
+  output.writeln('// GENERATED CODE - DO NOT MODIFY BY HAND\n');
+  output.writeln('class BuildConfig {');
+  output.writeln('  static const String kBuildTime = "${clock.now().toIso8601String()}";');
+  output.writeln('  static const int kBuildTimeMillisecondsSinceEpoch = ${clock.now().millisecondsSinceEpoch};');
+  output.writeln('  static const bool kDebug = ${buildInfo.isDebug};');
+  output.writeln('  static const String kModeName = "${buildInfo.modeName}";');
+  output.writeln('  static const String kFlavor = ${buildInfo.flavor != null ? "buildInfo.flavor" : null };');
+
+  // TODO (the4thfloor) add build-name and build-number from pubspec.yaml
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/buildconfig_generator.dart]: syntax for TODOs is `// TODO(username): description, https://link-to-issue

**Comment by Hixie** [comment]
I'll let @mravn-google or @tvolkert do the actual review, I'm not familiar the Android build system.

**Comment by tvolkert** [comment]
ping @mravn-google

**Comment by mravn-google** [comment]
Sorry for the latency. Current behavior is that following `flutter packages get`, both the `android/` and the `ios/` sub-projects can be built using platform-specific tooling (Gradle for Android, Cocoapods+Xcode for iOS) with no need for `flutter build xxx` or `flutter run`. Does this PR change that behavior?

**Comment by the4thfloor** [comment]
@mravn-google 

This behaviour (build/run the project without Flutter tooling) isn't changed. But maybe the `lib/build_config.g.dart` file isn't (re-)generated which will let the app crash if the Dart part tries to access it. I will check it.

**Comment by Hixie** [comment]
@the4thfloor Are you still looking into this? It's totally fine if not; we can take it over (though it might take longer). I'm just asking because I'm trying to clean out our review queue.

**Comment by the4thfloor** [comment]
@Hixie it is still in progress :-)
I'm waiting for https://github.com/flutter/flutter/pull/16857 to be merged because I want to put the `version` info into the generated `lib/build_config.g.dart` file

**Comment by mravn-google** [comment]
Seems to me Flutter tooling needs a design change to provide a good user experience for working with generated files. This PR makes generated files part of default operation, which worries me a bit.

What is the intended status of the generated file? Is it supposed to be checked into version control?
* If no, we'd want to add it to `.gitignore`, but then a newly checked-out project that uses `BuildConfig` in its Dart code will fail to be built by Gradle or Cocoapods+Xcode even after `flutter packages get` has been executed. That breaks a current property of Flutter tooling, as alluded to in an [earlier comment](https://github.com/flutter/flutter/pull/16934#issuecomment-388650240). 
* If yes, any call to `flutter run` will make the repo dirty. That does not seem right either.

Other questions:
* Should generated files be put in a special directory rather than next to `main.dart`?
* Should plugins be able to use the generated file? If so, how could one set up a Flutter plugin project to accomplish that (right now, we seem to end up with a two-way dependency between the plugin and the hosting/example app)?
* If somebody wanted to use protobufs or similar in their Flutter project, they would need generated Dart files too. One might wish for a standardized build step in the Flutter build tooling for codegen that should ideally work with codegen'ing plugins.

I think we should come up with reasonable answers to such questions, or at least make the generation of `BuildConfig` optional and non-default for now.

**Comment by the4thfloor** [comment]
This `BuildConfig` class is inspired by the Android version of this class. And on Android app development, we use this class often.
In an ideal world, there is a "hidden" `gen` directory where the build tool can put generated classes, but I didn't find a way to add a second source folder beside the `lib` folder.
A standardized build step in the Flutter build tooling would be great but I'm not sure if I'm the right person to architect this change (i guess there are lots of pitfalls, etc.).

For now, I will add a flag to enable/disable this generation step. I would prefer to add this flag as an option in the Flutter part in the pubspec.yaml file to enable/disable this project-wide for all runs/builds. Are you fine with this?

**Comment by mravn-google** [comment]
@devoncarew Any considerations on how the Flutter IDE plugins should handle codegen? This PR proposes to generate a `build_config.g.dart` file during execution of `flutter run` or `flutter build`.

**Comment by fmatosqg** [comment]
My 2 cents is leaving lib/ alone, but as said above there's no built in way to add a gen folder.

Maybe we add a new package to hold generated files as part of the build? The advantage is that devs will clearly know what code was typed (lib) and what was generated (myapp_gen).

**Comment by zoechi** [comment]
@fmatosqg all code-gen approaches write to `lib/` so I wouldn't be to worried about that.
I could imagine integrating https://github.com/dart-lang/build/tree/master/build_runner to be a great improvement and perhaps some config about what builders to run on what "lifecycle" events like before "flutter run", "flutter build", "packages get/upgrade", ...

**Comment by mravn-google** [comment]
Our work on add2app does not preclude code generation build steps. But for a seamless add2app experience, such steps should be executable via platform-specific build tooling (Gradle, CocoaPods and/or custom Xcode build scripts).

We already call `flutter build aot` from such tooling when building in non-debug mode and later `flutter build bundle` regardless of mode, so one approach would be extending `flutter_tools/gradle/flutter.gradle` and `flutter_tools/bin/xcode_backend.sh` with explicit calls to a Flutter CLI tool that does the code generation, before the optional `flutter build aot` step.

Maybe `flutter build precompile`?

**Comment by the4thfloor** [comment]
@alexlovar I guess the best way to get is to check out this branch (and maybe rebase it with master)

File: packages/flutter_tools/lib/src/buildconfig_generator.dart [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    output.writeln('  };');
+  }
+  output.writeln('}');
+
+  return output.toString();
+}
+
+dynamic _escape(dynamic value) {
+  if (value is String) {
+    return '\'$value\'';
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/buildconfig_generator.dart]: what is value contains an apostrophe or backslash?

File: packages/flutter_tools/lib/src/buildconfig_generator.dart [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    output.writeln('  static const Map<String, dynamic> kBuildArguments = {');
+    output.writeln('    ${args.options.map((String option) => '\'$option\': ${_escape(args[option])}').join(',\n    ')}');
+    output.writeln('  };');
+  }
+  output.writeln('}');
+
+  return output.toString();
+}
+
+dynamic _escape(dynamic value) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/buildconfig_generator.dart]: It's hard for me to understand the range and domain of this method. It takes any type, and returns either an escaped string or the type unmodified? What if the type is an object whose `toString` isn't valid Dart, say?

File: packages/flutter_tools/test/buildconfig_generator_test.dart [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+      const BuildInfo buildInfo = const BuildInfo(BuildMode.release, null, buildName: '1.0.2', buildNumber: 3);
+      final List<String> expected = <String>[
+        'static const bool kDebug = false;',
+        'static const String kModeName = \'release\';',
+        'static const String kVersionName = \'1.0.2\';',
+        'static const int kVersionNumber = 3;'
+      ];
+      await checkGenerator(manifest: manifest, buildInfo: buildInfo, expectedLines: expected);
+    });
+  });
```
Comment (Hixie) [inline packages/flutter_tools/test/buildconfig_generator_test.dart]: we should probably also have a test that verifies that the generated file compiles and analyzes cleanly.

File: packages/flutter_tools/test/buildconfig_generator_test.dart [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    Future<void> checkGenerator({
+      String manifest,
+      BuildInfo buildInfo,
+      List<String> expectedLines,
+    }) async {
+      final String projectPath = await createMinimalProject(manifest);
+
+      await generateBuildConfigClassAtPath(buildInfo: buildInfo, projectPath: projectPath);
+
+      final String srcFile = fs.path.join(projectPath, 'lib', 'build_config.g.dart');
```
Comment (Hixie) [inline packages/flutter_tools/test/buildconfig_generator_test.dart]: Would it make sense to put this in the `build/` directory?

**Comment by Hixie** [comment]
@the4thfloor Sorry that we are not being very responsive and helpful on this PR... the truth is that we don't really have any build system experts on the team, and so we're somewhat stumbling around in the dark on this topic. It doesn't help that we'd like to keep things consistent between iOS and Android, so ideally anyone who owns this area would be an expert it both. We're hoping to have a more coherent solution to codegen at some point in the future, something that works reliably with hot reload, for example, but we haven't yet started thinking about that much and so we're not really sure what we want this to look like. As @mravn-google points out above, this is also something that interacts with our existing-app integration (where the build is driven not from the `flutter` tool but from the OEM build system).

It might be best to file a bug and list all the requirements (some of which are in comments on this PR), and then have a discussion about the overall architecture that will address all these requirements, before landing this PR.

File: packages/flutter_tools/lib/src/buildconfig_generator.dart (hunk 6 above)
Comment (the4thfloor) [reply packages/flutter_tools/lib/src/buildconfig_generator.dart]: That is an excellent question. Currently, I don't know what happens, but I will look into it.

**Comment by the4thfloor** [comment]
@Hixie For me, it is okay when it needs longer than other PRs because it is a topic with a significant impact on the build system. 
And on the other hand, currently, I'm on an iOS project and have not much time for Flutter :-(

My lasts thoughts before I switch to iOS was to use the hidden .dart_tools folder to store the generated files. But Flutter doesn't use this folder. I have to look into it why Flutter doesn't use it.

But if there is a more significant change in the build system in the pipe, it doesn't make sense to add other complicated things. Maybe it is better to discuss it first in a bug ticket.

**Comment by ycktw** [comment]
Hi, Is the file path possible to be the same with flutter i18n ?
I found flutter i18n use the path to generate lib/generated/i18n.dart
and the code also easy to rename from build_config.g.dart to lib/generated/build_config.dart

I think the4thfloor did a good job, thanks again.

```
--- a/packages/flutter_tools/lib/src/buildconfig_generator.dart
+++ b/packages/flutter_tools/lib/src/buildconfig_generator.dart
@@ -16,7 +16,7 @@ import 'bundle.dart' as bundle;
 import 'flutter_manifest.dart';
 import 'globals.dart';

-const String _kFileName = 'build_config.g.dart';
+const String _kFileName = 'generated/build_config.dart';
```

**Comment by fmatosqg** [comment]
TL;DR I suggest using a convention where `*.g.dart` files get commited to git and `lib/generated/*.dart` are not.

I like `lib/generated/build_config.dart`, because it resonates with the idea of having `lib/generated` on `.gitignore` though not sure what's the role of that i18n file.

I believe 2 standards for generated files are needed, 1 for pre-build time such as autogenerated code that deals with json and people can argue whether they go in commits or note, and another for files that will be consistently overwritten during the build (as in C macros), meaning it's obvious nobody wants to commit them.

**Comment by Hixie** [comment]
fyi @matthew-carroll you may find this bug relevant to your interests.

I think we should close this PR for now, because we don't have a complete handle on the problem space (as per my last comment). I think the next step is to file an issue that describes the problem we want solved, and to discuss a design there.

**Comment by Hixie** [comment]
I'm going to close this PR per the last comment. Please do file an issue to discuss the design if you still desire a change in this area. Thanks.

**Comment by matthew-carroll** [comment]
Sounds good.  My 2 cents is that we need to figure out how to integrate Android's real BuildConfig on the engine side.  We probably shouldn't imitate something that already exists, and I think it's our own build system shortcoming that we can't offer this Class as it usually exists.

**Comment by SteveAlexander** [comment]
I was a little disappointed this didn't get merged, as I wanted to be able to read the app's version data from within my app.

I worked around it by adding `pubspec.yaml` as an asset in the `pubspec.yaml` file, then using the `pubspec_parse` package to parse it.

```
rootBundle.loadStructuredData<Pubspec>(
  'pubspec.yaml',
  (string) async => await Pubspec.parse(string)
);
```

https://pub.dartlang.org/packages/pubspec_parse

**Comment by the4thfloor** [comment]
@SteveAlexander try this plugin https://pub.dartlang.org/packages/package_info
